Cabal/src/Distribution/Simple/GHC/Build.hs

```diff
@@ -1,262 +1,137 @@
-        BS.pack $
-          escapeArgs $
-            extra_opts ++ renderGhcOptions comp platform (replOpts{ghcOptMode = NoFlag})
+{-
+Note [Build Target Dir vs Target Dir]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Where to place the build result (targetDir) and the build artifacts (buildTargetDir).
+
+\* For libraries, targetDir == buildTargetDir, where both the library and
+artifacts are put together.
+
+\* For executables or foreign libs, buildTargetDir == targetDir/<name-of-target-dir>-tmp, where
+    the targetDir is the location where the target (e.g. the executable) is written to
+    and buildTargetDir is where the compilation artifacts (e.g. Main.o) will live
+  Arguably, this difference should not exist (#9498) (TODO)
+
+For instance, for a component `cabal-benchmarks`:
+  targetDir == <buildDir>/cabal-benchmarks
+  buildTargetDir == <buildDir>/cabal-benchmarks/cabal-benchmarks-tmp
+
+Or, for a library `Cabal`:
+  targetDir == <buildDir>/.
+  buildTargetDir == targetDir
+
+Furthermore, we need to account for the limit of characters in ghc
+invocations that different OSes constrain us to. Cabal invocations can
+rapidly reach this limit, in part, due to the long length of cabal v2
+prefixes. To minimize the likelihood, we use
+`makeRelativeToCurrentDirectory` to shorten the paths used in invocations
+(see da6321bb).
+
+However, in executables, we don't do this. It seems that we don't need to do it
+for executable-like components because the linking step, instead of passing as
+an argument the path to each module, it simply passes the module name, the sources dir, and --make.
+RM: I believe we can use --make + module names instead of paths-to-objects
+for linking libraries too (2024-01) (TODO)
+-}
+
+-- | The main build phase of building a component.
+-- Includes building Haskell modules, extra build sources, and linking.
+build
+  :: Flag ParStrat
+  -> PackageDescription
+  -> BuildM ()
+build numJobs pkg_descr = do
+  verbosity <- buildVerbosity
+  component <- buildComponent
+  isLib <- buildIsLib
+  lbi <- buildLBI
+  clbi <- buildCLBI
+
+  -- Create a few directories for building the component
+  -- See Note [Build Target Dir vs Target Dir]
+  let targetDir_absolute = componentBuildDir lbi clbi
+      buildTargetDir_absolute
+        -- Libraries use the target dir for building (see above)
+        | isLib = targetDir_absolute
+        -- In other cases, use targetDir/<name-of-target-dir>-tmp
+        | targetDirName : _ <- reverse $ splitDirectories targetDir_absolute =
+            targetDir_absolute </> (targetDirName ++ "-tmp")
+        | otherwise = error "GHC.build: targetDir is empty"
+
+  liftIO $ do
+    createDirectoryIfMissingVerbose verbosity True targetDir_absolute
+    createDirectoryIfMissingVerbose verbosity True buildTargetDir_absolute
+
+  -- See Note [Build Target Dir vs Target Dir] as well
+  _targetDir <- makeRelativeToCurrentDirectory targetDir_absolute & liftIO
```

Comment:
The rest of the codebase uses `liftIO $ ...`. Could you switch to that?

Reply:
Yes.